### ISSUE TITLE
large memory usage

### ISSUE THREAD

**Issue by SalvatoreRa** [interpretml/interpret#397]
Hi,

I have a dataset with 10000 features and 500 examples (genomic data), I have used different models on that (XGBoost, neural network, and so on). However, when using EBM the kernel is dying, I have tried to execute the script in the minimal form from the terminal but it is stopped for memory usage. 

XGBoost takes around 40 seconds (on CPU) with the dataset, and while already downsized to 3000 features (just for testing) EBM takes 300 seconds. I tried to increment gradually the number of features to see when it stops and around 4000 it is using too much memory.

While reading the documentation and online I found that EBM is more computationally expensive than XGBoost, it seems to me that the memory usage here is too big. It is an error? there are some parameters I can use to reduce computational cost or memory usage?

sample code:

```
import random
from interpret import set_visualize_provider
from interpret.provider import InlineProvider
set_visualize_provider(InlineProvider())
import pandas as pd
from sklearn.model_selection import train_test_split
def random_dataset(examples = 500, features =10000):

    _X = np.random.rand(examples,features)
    _y = []
    for i in range(0,examples):
        n = random.randint(0,1)
        _y.append(n)

    _X = pd.DataFrame(_X, columns = ["feature" +str(i)  for i in range(1,_X.shape[1] +1,1)],
                     index = ["example" +str(i) for i in range(1,_X.shape[0] +1,1)])
    
    return _X, _y

X, y = random_dataset(examples = 500, features =10000)
seed = 1
X_train, X_test, y_train, y_test = train_test_split(X, Y, 
                                                    test_size=0.20, random_state=seed)
ebm = ExplainableBoostingClassifier(random_state=seed)
ebm.fit(X_train, y_train)
print(ebm.score(X_test,  y_test))

ebm_global = ebm.explain_global()

jsonStr = ebm_global.data()
feat_imp = pd.DataFrame(columns = ['names', 'scores'])
feat_imp['names'] =jsonStr['names']
feat_imp['scores'] =jsonStr['scores']
```
thank you for your help

**Comment by paulbkoch** [comment]
Hi @SalvatoreRa -- Very likely the issue here is caused by the interaction detection algorithm which by default evaluates all possible pairs.  I would recommend setting interactions to 0, which will turn off the default behavior of searching through all possible pairs.  You can still specify exact pairs, and it should work without issues then.  With 500 examples, I suspect pairs won't be a strong factor, but if you really want to try adding pairs back in we now have a separate utility function that can evaluate pairs which you could use to sidestep the memory issues, and also potentially the computational issues if you're ok with applying a heuristic to eliminate many of the possible pairs.

**Comment by SalvatoreRa** [comment]
thank you for your answer, I am strongly interested in the interpretability side, does shutting off the interaction affects the interpretability?

**Comment by paulbkoch** [comment]
Hi @SalvatoreRa -- Most people would argue that EBMs without interactions are slightly more interpretable.  Whether to include pair interactions or not, and how many, is a tradeoff that you need to consider when deciding if you want to optimize for interpretability or accuracy.

Again though, with 500 samples I do not think you have enough data to find meaningful interactions, so I'd probably turn them off anyways.
